src/main/kotlin/org/stellar/walletsdk/Anchor.kt

```diff
@@ -0,0 +1,172 @@
+package org.stellar.walletsdk
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import org.stellar.sdk.Network
+import org.stellar.sdk.Server
+import org.stellar.walletsdk.util.GsonUtils
+import org.stellar.walletsdk.util.StellarToml
+import org.stellar.walletsdk.util.interactiveFlow
+
+/**
+ * Build on/off ramps with anchors.
+ *
+ * @property server Horizon [Server] instance
+ * @property network Stellar [Network] instance
+ * @property httpClient optional custom HTTP client, uses [OkHttpClient] by default
+ */
+class Anchor(
+  private val server: Server,
+  private val network: Network,
+  private val httpClient: OkHttpClient = OkHttpClient()
+) {
+  /**
+   * Get anchor information from a TOML file.
+   *
+   * @param assetIssuer Stellar address of the asset issuer (anchor)
+   *
+   * @return TOML file content
+   */
+  suspend fun getInfo(assetIssuer: String): Map<String, Any> {
+    return CoroutineScope(Dispatchers.IO)
+      .async {
+        val toml = StellarToml(stellarAddress = assetIssuer, server, httpClient)
+
+        return@async toml.getToml()
+      }
+      .await()
+  }
+
+  /**
+   * Available anchor services and information about them. For example, limits, currency, fees,
+   * payment methods.
+   *
+   * @param serviceUrl URL where `/info` endpoint is hosted
+   *
+   * @return a list of available anchor services
+   *
+   * @throws [NetworkRequestFailedException] if network request fails
+   * @throws [InvalidAnchorServiceUrl] if provided service URL is not a valid URL
+   */
+  suspend fun getServicesInfo(serviceUrl: String): AnchorServiceInfo {
+    try {
+      val url = serviceUrl.toHttpUrl()
+      val urlBuilder = HttpUrl.Builder().scheme(url.scheme).host(url.host).port(url.port)
+
+      url.pathSegments.forEach { urlBuilder.addPathSegment(it) }
+      urlBuilder.addPathSegment("info")
+
+      val infoUrl = urlBuilder.build().toString()
+
+      val request = Request.Builder().url(infoUrl).build()
+      val gson = GsonUtils.instance!!
+
+      return CoroutineScope(Dispatchers.IO)
+        .async {
+          httpClient.newCall(request).execute().use { response ->
+            if (!response.isSuccessful) throw NetworkRequestFailedException(response)
+
+            val infoResponse = response.body!!.charStream()
+            return@async gson.fromJson(infoResponse, AnchorServiceInfo::class.java)
+          }
+        }
+        .await()
+    } catch (e: Exception) {
+      throw InvalidAnchorServiceUrl(e)
+    }
```

Comment:
We should really try and avoid wrapping large chunks of code in try blocks. Where is the line of code that would raise an exception resulting from an invalid URL?

Reply:
It's *only* 22 lines above. :wink: I'll update to move the exception closer to the source.